rfcs/42-world-cell-queries.md

```diff
@@ -0,0 +1,236 @@
+
+Allow performing world modifications through WorldCell shared access using command API.
+
+Make the world cell queries always reflect the most recent updated world data, without requiring explicit
+sync points inside the single-threaded systems.
+
+## Motivation
+
+Bevy ECS is currently quite well suited for high data throughput and relatively low complexity scenarios.
+Most focus is being put on general performance and ergonomics of safe multi-threaded query system.
+That high performance API imposes some hard limitations on how to structure the data, effectively requiring
+every system to be carefully designed around rust borrowing rules, as well as making complex data dependency chains
+sometimes very hard to deal with.
+
+While this trade-off might be right for a lot of scenarios, there is currently no way out of it. Most notably, it can
+easily get in the way of complex gameplay logic prototyping, where development speed is much more important than raw performance.
+Similarly, some systems might not be performance critical, but annoying to write with usual query limitations in place.
+
+The trade-off of performance versus code complexity should be a decision that a developer could make on their own.
+
+## User-facing explanation
+
+### World cell query
+
+When having access to a `WorldCell`, you are free to issue any ECS queries you wish, at any time. Even ones that seemingly break rust principles.
+
+```rust
+pub fn single_threded_system(world: &mut World) {
+    let mut cell = world.cell();
+
+    // create two seemingly conflicting queries and iterate over them
+    let query_ref = cell.query::<(Entity, &Transform)>();
+    let query_mut = cell.query::<(Entity, &mut Transform)>();
+    let mut iter_ref = query_ref.iter();
+    let mut iter_mut = query_mut.iter();
+
+    iter_ref.next(); // offset mutable query by one item to avoid aliasing
+
+    for (previous, next) in iter_mut.zip(iter_ref) {
+        // assign a transform from one entity to another
+        *previous.1 = next.1.clone();
+    }   
+}
+```
+
+In the above example, the two queries would not be allowed to coexist in a single system. But they are still used in a way that will never lead to mutable aliasing.
+This is because one of the queries is being offset during actual iteration. Because the rust aliasing invariant checks are moved to runtime, this code is able to run.
+
+Any unexpected situation when the safety invariant is not upheld is being detected at runtime, at the time of access to given entity.
+In above example, removing `iter_mut.next()` call would lead to a runtime panic: 
+```
+Component 'Transform' of entity 1234v5 already borrowed
+```
+That panic is being issued from inside the query iterator's `next` method, which is implcitly called by the `for` loop.
+
+Any combination of queries is valid and both can be iterated over at the same time. It's the code author's responsibility to check if mutability rules are not violated.
```

Comment:
I can understand the justification for this, but I'm very nervous about encouraging users to do this, even for prototype code. It introduces a lot of potential for spooky-action-at-a-distance bugs, where the game will just panic at runtime in strange edge cases. Archetype invariants would help move those panics to the point of forbidden insertion, which would help, but getting those working in a performant fashion is a major endeavor.

If this feature was added to Bevy, I would write a lint forbidding its use in any of our company code bases due to the risk involved.